### PR TITLE
Bind SD layer to SOMEIP layer

### PR DESCRIPTION
**Pull request by kruemelbacke**
Solve the issue #2343

### CODE REVIEW

**Comment by polybassa** [comment]
Thanks for this PR. Could you please provide an unit-test.

File: scapy/contrib/automotive/someip.py [hunk 1]
```diff
@@ -213,6 +214,18 @@ def fragment(self, fragsize=1392):
...
         return lst
 
+    def guess_payload_class(self, payload):
+        """
+        Defines if next payload is SD layer or default layer   
+        :param str payload: the layer's payload
+        :return: the payload class
+        """
+        if self.getfieldval("msg_id").srv_id == 0xffff:
+            self.show()
```
Comment (polybassa) [inline scapy/contrib/automotive/someip.py]: you shouldn't call `show` here

File: scapy/contrib/automotive/someip.py [hunk 2]
```diff
@@ -213,6 +214,18 @@ def fragment(self, fragsize=1392):
...
+            print(payload)
+            return SD(payload)
```
Comment (polybassa) [inline scapy/contrib/automotive/someip.py]: I think you have to return just `SD`

File: scapy/contrib/automotive/someip.py (hunk 1 above)
Comment (kruemelbacke) [reply scapy/contrib/automotive/someip.py]: oops, I forgot to remove my test prints

**Comment by kruemelbacke** [comment]
Need help to finish my work: I merged 'someip.py' and 'someip_sd.py' and deleted 'someip_sd.py'. Some builds don't work now...

**Comment by polybassa** [comment]
I think you have to remove the someip_sd.py file. It looks like the file is still present, but empty

File: scapy/contrib/automotive/someip.py [hunk 3]
```diff
@@ -1,3 +1,5 @@
+#! /usr/bin/env python
```
Comment (guedou) [inline scapy/contrib/automotive/someip.py]: This change is not needed, please revert it.

File: scapy/contrib/automotive/someip.py [hunk 4]
```diff
@@ -213,6 +220,13 @@ def fragment(self, fragsize=1392):
 
         return lst
 
+    def guess_payload_class(self, payload):
+        """Defines if next payload is SD layer or default layer"""
+        if self.getfieldval("msg_id").srv_id == 0xffff:
```
Comment (guedou) [inline scapy/contrib/automotive/someip.py]: Can you use `self.msg_id.srv_id` instead?

File: scapy/contrib/automotive/someip.py [hunk 5]
```diff
@@ -221,3 +235,310 @@ def _bind_someip_layers():
...
+        _SDOption_Header,
+        IPField("addr", "0.0.0.0"),
+        _SDOption_Tail
+    ]
+
+
+class _SDOption_IP6(_SDOption):
+    fields_desc = [
+        _SDOption_Header,
+        IP6Field("addr", "2001:cdba:0000:0000:0000:0000:3257:9652"),
```
Comment (guedou) [inline scapy/contrib/automotive/someip.py]: Is this address specific to SD? If not, I will suggest to use `2001:db8::1` instead.

File: scapy/contrib/automotive/someip.py [hunk 6]
```diff
@@ -221,3 +235,310 @@ def _bind_someip_layers():
...
+        if name in self.FLAGSDEF:
+            return ((self.flags & self.FLAGSDEF[name].mask) >>
+                    self.FLAGSDEF[name].offset)
+        else:
+            return None
+
+    def set_flag(self, name, value):
+        name = name.upper()
+        if name in self.FLAGSDEF:
+            self.flags = (self.flags &
```
Comment (guedou) [inline scapy/contrib/automotive/someip.py]: Can you split this on different lines?
For example
```
self.flags &= ctypes.c_ubyte(~self.FLAGSDEF[name].mask).value)
self.flags |= (value & 0x01) << self.FLAGSDEF[name].offset
```

**Comment by kruemelbacke** [comment]
@guedou 
> Thanks for your PR. Please apply my comments and add unit tests.

First, I'm sorry for my bad work. I'm new to github and also nearly new to python. 
Second, how can I add unit tests?

> Also, is there a reason why the `someip_sd.py` was deleted?

The reason why I deleted some_ip is that when you use the someip contrib, you usually need also the someip_sd contrib. In my opinion a basic function (Service Discovery) of a protocol should be already implemented in the base contrib. There is no need to split these layers into different files. For example, this was my use case: I read in a pcap file holding SOME/IP packets and I didn't understand the payload of most packets until I realized these were SD packets.

File: scapy/contrib/automotive/someip.py (hunk 5 above)
Comment (kruemelbacke) [reply scapy/contrib/automotive/someip.py]: I didn't changed any classes of someip-sd. But this looks good.

File: scapy/contrib/automotive/someip.py (hunk 4 above)
Comment (kruemelbacke) [reply scapy/contrib/automotive/someip.py]: Yes, sure

File: scapy/contrib/automotive/someip.py (hunk 6 above)
Comment (kruemelbacke) [reply scapy/contrib/automotive/someip.py]: I didn't write this part of code myself and I don't really understand what it does. If you are sure these two lines will work I will replace the actual line.

**Comment by polybassa** [comment]
maybe a rebase to the current master fixes your travis tests

**Comment by kruemelbacke** [comment]
> 
> 
> maybe a rebase to the current master fixes your travis tests

I'm new to git, I don't know how to do this. Could you explain what you mean?

**Comment by polybassa** [comment]
https://git-scm.com/book/de/v2/Git-Branching-Rebasing
```
git pull -r <origin> master
```
execute it on your branch with your secdev/scapy as origin

**Comment by polybassa** [comment]
Could you try to rebase to the latest version of scapys master branch

**Comment by polybassa** [comment]
I think this PR can be closed, since #2441 and #2443 are made.

**Comment by kruemelbacke** [comment]
Sorry for not finishing my work, I hope I could still contribute to progress. Thanks for taking over the Pull Request.

**Pull request closed without merging**